run/runner.go

```diff
@@ -1,21 +1,46 @@
 package run
 
-import (
-	"github.com/nextmv-io/sdk/run/decode"
-	"github.com/nextmv-io/sdk/run/encode"
-)
+import "context"
 
 // Runner defines the interface of the runner.
-type Runner interface {
-	// Run invokes a solver by invoking the associated handler.
-	Run()
+type Runner[RunnerConfig, Input, Option, Solution any] interface {
+	// Run runs the runner.
+	Run(context.Context) error
+	// SetIOProducer sets the ioProducer of a runner.
+	SetIOProducer(IOProducer[RunnerConfig])
+	// SetInputDecoder sets the inputDecoder of a runner.
+	SetInputDecoder(Decoder[Input])
+	// SetOptionDecoder sets the optionDecoder of a runner.
+	SetOptionDecoder(Decoder[Option])
+	// SetAlgorithm sets the algorithm of a runner.
+	SetAlgorithm(Algorithm[Input, Option, Solution])
+	// SetEncoder sets the encoder of a runner.
+	SetEncoder(Encoder[Solution, Option])
+	// GetEncoder returns the encoder of a runner.
+	GetEncoder() Encoder[Solution, Option]
+	// RunnerConfig returns the runnerConfig of a runner.
+	RunnerConfig() RunnerConfig
+}
+
+// IOProducer is a function that produces the input, option and writer.
+type IOProducer[RunnerConfig any] func(context.Context, RunnerConfig) IOData
+
+// Decoder is a function that decodes a reader into a struct.
+type Decoder[Input any] func(context.Context, any) (Input, error)
```

Comment:
Because it doesn't have to be an `io.Reader`. There is no limit as to what an `IOProducer` can provide as the "data source". The `Decoder` will have to be able to handle it though (see comment below).